In [ ]:
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

In [ ]:
df = pd.read_excel('/train_pysics_che_bio.xlsx')
df.head()

Id                                            Comment      Topic
0   0x840  A few things. You might have negative- frequen...    Biology
1   0xbf0  Is it so hard to believe that there exist part...    Physics
2  0x1dfc                                     There are bees    Biology
3   0xc7e  I'm a medication technician. And that's alot o...    Biology
4   0xbba                     Cesium is such a pretty metal.  Chemistry

In [ ]:
df.isnull().sum()

Id         0
Comment    2
Topic      0
dtype: int64

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(df['Comment'], df['Topic'], test_size=0.2, random_state=42)
X_train.shape

(6956,)

kita bagi menjadi train dan test data, dengan ratio 80:20

In [ ]:
vectorizer = TfidfVectorizer()

define vectorizer kita yaitu TFIDF

In [ ]:
X_train_tfidf = vectorizer.fit_transform(X_train.values.astype('U'))
X_test_tfidf = vectorizer.transform(X_test.values.astype('U'))

fit data predictor ke tfidf sehingga menghasilkan output data train dalam bentuk vector, namun jika hanya menggunakan data train dalam bentuk awal nya yaitu X_train dan X_test maka kita akan dapat error np.nan, oleh karena itu kita perlu mengubah nya menjadi unicode agar dapat dibaca oleh tfidf

In [ ]:
model_nb = MultinomialNB()
model_nb.fit(X_train_tfidf, y_train)
predictions_nb = model_nb.predict(X_test_tfidf)
accuracy_nb = accuracy_score(y_test, predictions_nb)
print("Accuracy (Multinomial Naive Bayes):", accuracy_nb)

Accuracy (Multinomial Naive Bayes): 0.6831512363427257


In [ ]:
model_svm = SVC()
model_svm.fit(X_train_tfidf, y_train)
predictions_svm = model_svm.predict(X_test_tfidf)
accuracy_svm = accuracy_score(y_test, predictions_svm)
print("Accuracy (Support Vector Machine):", accuracy_svm)

Accuracy (Support Vector Machine): 0.730879815986199


Hasil dari kedua model tersebut mendekati dimana jika kita hanya bandingkan accuracy yg mana Multinomial Naive Bayes memiliki 0.68 dan SVM memiliki 0.73,